src/network/rpc/client.rs

```diff
@@ -690,43 +390,62 @@ impl Client {
 	}
 
 	pub async fn get_validator_set_at(&self, block_hash: H256) -> Result<Option<Vec<AccountId32>>> {
-		self.execute_sync(|response_sender| {
-			Box::new(GetValidatorSetAt {
-				block_hash,
-				response_sender: Some(response_sender),
+		let res = self
+			.with_retries(|client| {
+				let validators_key = api::storage().session().validators();
+				async move { client.storage().at(block_hash).fetch(&validators_key).await }
 			})
-		})
-		.await
+			.await
+			.map_err(Report::from)?;
+
+		Ok(res)
 	}
 
-	pub async fn submit_signed_and_watch(
+	pub async fn submit_signed_and_wait_for_finalized<Call: subxt::tx::TxPayload>(
 		&self,
-		extrinsic: Payload<SubmitData>,
-		pair_signer: PairSigner<AvailConfig, Pair>,
-		params: AvailExtrinsicParams,
-	) -> Result<TxProgress<AvailConfig, OnlineClient<AvailConfig>>> {
-		self.execute_sync(|response_sender| {
-			Box::new(SubmitSignedAndWatch {
-				extrinsic,
-				pair_signer,
-				params,
-				response_sender: Some(response_sender),
+		call: &Call,
+		signer: &PairSigner<AvailConfig, Pair>,
+		other_params: avail_subxt::primitives::AvailExtrinsicParams,
+	) -> Result<subxt::blocks::ExtrinsicEvents<AvailConfig>> {
+		let tx_progress = self
+			.with_retries(|client| {
+				let other_params = other_params.clone();
+				async move {
+					client
+						.tx()
+						.sign_and_submit_then_watch(call, signer, other_params)
```

Comment:
Did we consider that there might be multiple undesired submtis if we do retries here?